File: src/main/domain/stock/Quote.java
package domain.stock;

import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;

//===========================================================//
/**
 * Represents the latest market quote for a stock symbol.
 * <p>
 *    Stores readable version of data after it has been received
 *    by an external Provider.
 * </p>
 *
 * <p>
 *    Example:
 *       Symbol = "NET" means CloudFare stock.
 * </p>
 *
 * @param symbol
 * @param currentPrice
 * @param change
 * @param percentChange
 * @param highPrice
 * @param lowPrice
 * @param openPrice
 * @param prevClosePrice
 */
//===========================================================//

public record Quote(
  String symbol,
  double currentPrice,
  double change,
  double percentChange,
  double highPrice,
  double lowPrice,
  double openPrice,
  double prevClosePrice,
  long receivedAtMillis
) {

   private static final DateTimeFormatter FORMATTER =
     DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
       .withZone(ZoneId.systemDefault());


   public Quote{
      if (symbol == null || symbol.isBlank()) throw new IllegalArgumentException("Symbol is missing");
      if (currentPrice <= 0d) throw new IllegalArgumentException("Current price is invalid");
      if (receivedAtMillis <= 0L) throw new IllegalArgumentException("Received at is invalid");
   }

   public String getFormattedReceivedAt(){
      return FORMATTER.format(Instant.ofEpochMilli(receivedAtMillis));
   }
}

File: src/main/kotlin/domain/stock/Quote.kt
package domain.stock

import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

//===========================================================//
/**
 * Represents the latest market quote for a stock symbol.
 * 
 * 
 * Stores readable version of data after it has been received
 * by an external Provider.
 * 
 * 
 * 
 * 
 * Example:
 * Symbol = "NET" means CloudFare stock.
 * 
 * 
 * @param symbol
 * @param currentPrice
 * @param change
 * @param percentChange
 * @param highPrice
 * @param lowPrice
 * @param openPrice
 * @param prevClosePrice
 */
//===========================================================//

data class Quote(
    val symbol: String,
    val currentPrice: Double,
    val change: Double,
    val percentChange: Double,
    val highPrice: Double,
    val lowPrice: Double,
    val openPrice: Double,
    val prevClosePrice: Double,
    val receivedAtMillis: Long
) {
    val formattedReceivedAt = s_Formatter.format(Instant.ofEpochMilli(receivedAtMillis))

    companion object {
        private val s_Formatter: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
            .withZone(ZoneId.systemDefault())
    }

    init {
        require(!symbol.isBlank()) { "Symbol is missing" }
        require(currentPrice >= 0.0) { "Current price is invalid" }
        require(receivedAtMillis > 0L) { "Received at is invalid" }
    }
}
